### DISCUSSION TITLE
Question about how updates with nested createMany handles concurrency

### DISCUSSION THREAD

**Discussion by heyskylark** [prisma/prisma#23351] [Q&A] [answered]
### Question

Hey all, I have a quick question about how update constructs queries, under heavier load where the same resource gets updated multiple times in quick succession I've noticed I get this error:

```
"An operation failed because it depends on one or more records that were required but not found. No 'X' record (needed to inline the relation on 'Y' record) was found for a nested createMany on relation 'YToXRecord'.
```

We also currently use Serializable transaction mode so part of me has a hunch that it's possible that might be a part of the issue as well

Essentialy, I have a prisma query like so:
```js
 prisma.recordX.update({
  where: {
    id: <record-to-update-id>,
    lastProcessedAtDate: {
      lt: lastProcessedAt,
    },
  },
  data: {
    somRow: "updated data",
    recordY: {
      deleteMany: {},
      createMany: {
        data: [array of new record Y's],
      },
    },
  },
}),
```

During larger loads sometimes, the same recordX gets updated multiple times in parallel, which appears to be causing the createMany of record Y to fail. One option is I can probably batch same userIds in our queues better to avoid this more, but was curious about getting more into the root cause of what might cause something like this first. As well as how prisma goes about constructing the queries for an update that has batch queries nested within it? The error seems to imply the parent record does not exist, which I'm having trouble understanding since that record is only ever updated & never deleted.

### How to reproduce (optional)

Run a prisma query like:
```js
prisma.recordX.update({
  where: {
    id: <record-to-update-id>,
    lastProcessedAtDate: {
      lt: lastProcessedAt,
    },
  },
  data: {
    somRow: "updated data",
    recordY: {
      deleteMany: {},
      createMany: {
        data: [array of new record Y's],
      },
    },
  },
}),
```

Multiple times at the same time with the same record.

### Expected behavior (optional)

I expect that the update just occurs twice, instead of one of the parallel updates failing with the above error.

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
model UserTieredEmblem {
  id                  String   @id @default(uuid()) @db.Uuid
  userId              String   @unique @map("user_id") @db.Uuid
  totalPoints         Int      @map("total_points")
  blockchainLogId     String?  @map("blockchain_log_id") @db.Uuid
  lastProcessedAtDate DateTime @default(now()) @map("last_processed_at_date") @db.Timestamptz(6)

  createdAt DateTime @default(now()) @map("created_at") @db.Timestamptz(6)
  updatedAt DateTime @default(now()) @updatedAt @map("updated_at") @db.Timestamptz(6)

  user         User          @relation(fields: [userId], references: [id], onDelete: Cascade, onUpdate: NoAction)
  emblemScores EmblemScore[]

  @@index([lastProcessedAtDate], map: "user_tiered_emblems_last_processed_at")
  @@index([totalPoints], map: "user_tiered_emblems_total_points")
  @@index([updatedAt], map: "user_tiered_emblems_updated_at")
  @@map("user_tiered_emblems")
}

// Used to see (1) what emblems a user has and (2) the amount this emblem contributes to the user's tiered emblem score
model EmblemScore {
  id                 String @id @default(uuid()) @db.Uuid
  userTieredEmblemId String @map("user_tiered_emblem_id") @db.Uuid

  emblem String

  // extra points applied for boostable emblems (e.g. if you have an Azuki that has 3 lightning traits, it gets a boost for the lightning emblem)
  boostPoints Int @default(0) @map("boost_points")

  createdAt DateTime @default(now()) @map("created_at") @db.Timestamptz(6)
  updatedAt DateTime @default(now()) @updatedAt @map("updated_at") @db.Timestamptz(6)

  userTieredEmblem UserTieredEmblem @relation(fields: [userTieredEmblemId], references: [id], onDelete: Cascade, onUpdate: NoAction)

  // json that holds the qualifiers for the emblem (the point contribution for each, the qualifying tokens if applicable, etc)
  qualifiers Json

  @@unique([emblem, userTieredEmblemId])
  @@map("emblem_scores")
}

**Comment by ludralph** [comment, marked as answer]
Hi @heyskylark 👋 

Thank you for raising this question.

You're performing multiple concurrent update operations on the same `recordX`. This could lead to a race condition where two or more operations do not find the record and therefore try to create that record. In this situation, one of the operations successfully creates the new record but the other operations fail and returns the error you encountered.

Nested update queries are only available in the context of a top-level update query. If the parent record does not exist, Prisma Client throws an exception. If the related record that you want to update does not exist, Prisma Client also throws an exception ([see the documentation](https://www.prisma.io/docs/orm/reference/prisma-client-reference#update-1)).

To avoid this issue, you could consider using nested writes feature supported by create and update in Prisma Client. If any operation fails, Prisma Client rolls back the entire transaction. Nested writes are not currently supported by top-level bulk operations like `deleteMany` and `updateMany` ([see the documentation](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#nested-writes-1)).

However, if you're dealing with a large number of concurrent updates, you might need to implement some form of queue or batching system to ensure that updates to the same record are not attempted simultaneously.

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise, please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by heyskylark** [reply]
Hey this was exactly what I needed, thank you so much for the response!
